muon/_core/preproc.py

```diff
@@ -720,19 +720,27 @@ def func(x):
         else:
             raise ValueError("When providing obs_names directly, func has to be None.")
 
-    # Subset .obs
-    data._obs = data.obs[obs_subset]
-    data._n_obs = data.obs.shape[0]
-
-    # Subset .obsm
-    for k, v in data.obsm.items():
-        data.obsm[k] = v[obs_subset]
-
-    # Subset .obsp
-    for k, v in data.obsp.items():
-        data.obsp[k] = v[obs_subset][:, obs_subset]
-
     if isinstance(data, AnnData):
+        # Collect elements to subset
+        # NOTE: accessing them after subsetting .obs
+        # will fail due to _validate_value()
+        obsm = dict(data.obsm)
```

Comment:
probably need to use `_remove_unused_categories` on the dataframe - please open an issue in anndata about making this public